### PR TITLE
Solution DAO-4 was added firstly

### PR DESCRIPTION
**Pull request by tbondare**

### CODE REVIEW

**Review by YuraBondarenko — APPROVED** [review]

File: src/main/java/taxiapp/Main.java [hunk 1]
```diff
@@ -22,5 +26,40 @@ public static void main(String[] args) {
...
+        DriverService driverService = (DriverService)
+                injector.getInstance(DriverService.class);
+        Driver driver1 = new Driver("Alex", "licenseNumber-27");
+        Driver driver2 = new Driver("Antonio", "licenseNumber-35");
+        driverService.create(driver1);
+        driverService.create(driver2);
+        System.out.println(driverService.get(1L));
+        System.out.println(driverService.getAll().toString());
+        Driver driver3 = new Driver("Sancho", "licenseNumber-389");
+        driver3.setId(2L);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/Main.java:39]: this is not a good practice to change the id

the example of using the `update()` method can be the following:

```
        Driver driver1 = new Driver("Alex", "licenseNumber-27");
        Driver driver2 = new Driver("Antonio", "licenseNumber-35");
        driverService.create(driver1);
        driverService.create(driver2);
        Driver antonio = driverService.get(2L);
        antonio.setLicenseNumber("007");
        
        driverService.update(antonio);
```

you change Antonio locally and then call `update()` to save these changes in the DB

File: src/main/java/taxiapp/Main.java [hunk 2]
```diff
@@ -22,5 +26,40 @@ public static void main(String[] args) {
...
+
+        CarService carService = (CarService)
+                injector.getInstance(CarService.class);
+        Car reno = new Car("5a", manufacturerReno);
+        Car bmw = new Car("x6", manufacturerBmw);
+        carService.create(reno);
+        carService.create(bmw);
+        reno.addToDriveListById(driver1.getId());
+        reno.addToDriveListById(driver2.getId());
+        bmw.addToDriveListById(driver3.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/Main.java:53]: you should have the following methods in CarService, not in the Car:

- void addDriverToCar(Driver driver, Car car);
- void removeDriverFromCar(Driver driver, Car car);

File: src/main/java/taxiapp/dao/CarDao.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package taxiapp.dao;
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Car;
+
+public interface CarDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDao.java:8]: redundant empty line

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+import java.util.Optional;
+import taxiapp.db.Storage;
+import taxiapp.lib.Dao;
+import taxiapp.model.Car;
+import taxiapp.model.Driver;
+
+@Dao
+public class CarDaoImpl implements CarDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:13]: same

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public Car create(Car car) {
+        Storage.addCar(car);
+        return Storage.cars.get(Storage.cars.size() - 1);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:17]:
```suggestion
        return car;
```

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public List<Car> getAll() {
+        return Storage.cars;
+    }
+
+    @Override
+    public Car update(Car car) {
+        Car oldValue = get(car.getId()).orElseThrow();
+        Storage.cars.remove(oldValue);
+        Storage.cars.add(car);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:36]: this way the car gets the new id after update
this is not right

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        return oldValue;
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:38]:
```suggestion
    public Car update(Car car) {
        IntStream.range(0, Storage.cars.size())
                .filter(i -> Storage.cars.get(i).getId().equals(car.getId()))
                .forEach(i -> Storage.cars.set(i, car));
        return car;
    }
```
let's replace each occurrence of the car with the same id in Storage - with the new car

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.cars.remove(get(id).orElseThrow());
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:42]: use `removeIf()` method

File: src/main/java/taxiapp/dao/CarDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        List<Car> cars = new ArrayList<>();
+        for (Car car : Storage.cars) {
+            for (Driver driver : car.getDriverList()) {
+                if (driver.getId().equals(driverId)) {
+                    cars.add(car);
+                    break;
+                }
+            }
+        }
+        return cars;
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/CarDaoImpl.java:56]: let's use Stream API here instead
you can call `getAll()` method here to get all the cars

File: src/main/java/taxiapp/dao/DriverDao.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package taxiapp.dao;
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Driver;
+
+public interface DriverDao {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/DriverDao.java:8]: redundant empty line

File: src/main/java/taxiapp/dao/DriverDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import taxiapp.db.Storage;
+import taxiapp.lib.Dao;
+import taxiapp.model.Driver;
+
+@Dao
+public class DriverDaoImpl implements DriverDao {
+    @Override
+    public Driver create(Driver driver) {
+        Storage.addDriver(driver);
+        return Storage.drivers.get(Storage.drivers.size() - 1);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/DriverDaoImpl.java:14]:
```suggestion
        return driver;
```

File: src/main/java/taxiapp/dao/DriverDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    public List<Driver> getAll() {
+        return Storage.drivers;
+    }
+
+    @Override
+    public Driver update(Driver driver) {
+        Driver oldValue = get(driver.getId()).orElseThrow();
+        Storage.drivers.remove(oldValue);
+        Storage.drivers.add(driver);
+        return oldValue;
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/DriverDaoImpl.java:34]: same comment as for the CarDaoImpl

File: src/main/java/taxiapp/dao/DriverDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.drivers.remove(get(id).orElseThrow());
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/dao/DriverDaoImpl.java:39]: same

File: src/main/java/taxiapp/service/DriverService.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Driver;
+
+public interface DriverService {
+
+    Driver create(Driver driver);
+
+    Optional<Driver> get(Long id);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/DriverService.java:11]: we should return Driver on the service level, not an Optional

File: src/main/java/taxiapp/service/DriverService.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
+package taxiapp.service;
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Driver;
+
+public interface DriverService {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/DriverService.java:8]: redundant empty line

File: src/main/java/taxiapp/service/DriverServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.List;
+import java.util.Optional;
+import taxiapp.dao.DriverDao;
+import taxiapp.lib.Inject;
+import taxiapp.lib.Service;
+import taxiapp.model.Driver;
+
+@Service
+public class DriverServiceImpl implements DriverService {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/DriverServiceImpl.java:12]: same

File: src/main/java/taxiapp/service/DriverServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    @Inject
+    DriverDao driverDao;
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/DriverServiceImpl.java:14]: what is missing here?

File: src/main/java/taxiapp/service/CarService.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
+package taxiapp.service;
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Car;
+import taxiapp.model.Driver;
+
+public interface CarService {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarService.java:9]: redundant empty line

File: src/main/java/taxiapp/service/CarService.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    Car create(Car car);
+
+    Optional<Car> get(Long id);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarService.java:12]:
```suggestion
    Car get(Long id);
```

File: src/main/java/taxiapp/service/CarServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+import taxiapp.dao.CarDao;
+import taxiapp.dao.DriverDao;
+import taxiapp.lib.Inject;
+import taxiapp.lib.Service;
+import taxiapp.model.Car;
+import taxiapp.model.Driver;
+
+@Service
+public class CarServiceImpl implements CarService {
+
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarServiceImpl.java:14]: redundant empty line

File: src/main/java/taxiapp/service/CarServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    @Inject
+    DriverDao driverDao;
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarServiceImpl.java:16]: let's remove this

File: src/main/java/taxiapp/service/CarServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Inject
+    CarDao carDao;
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarServiceImpl.java:19]: add access modifier

File: src/main/java/taxiapp/service/CarServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return carDao.delete(id);
+    }
+
+    @Override
+    public void addDriverToCar(Driver driver, Car car) {
+        car.getDriverList().add(driver);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarServiceImpl.java:48]: let's call `carDao.update(car)` after we added or removed the driver

File: src/main/java/taxiapp/service/CarServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        car.getDriverList().remove(driver);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/service/CarServiceImpl.java:53]: same

File: src/main/java/taxiapp/model/Car.java [hunk 25]
```diff
@@ -33,4 +42,18 @@ public void setModel(String model) {
     public void setManufacturer(Manufacturer manufacturer) {
         this.manufacturer = manufacturer;
     }
+
+    public void addToDriveListById(Long driverId) {
+        driverList = Storage.drivers.stream()
+                        .filter(d -> d.getId().equals(driverId))
+                        .collect(Collectors.toList());
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/model/Car.java:50]: let's remove this method

File: src/main/java/taxiapp/model/Car.java [hunk 26]
```diff
@@ -33,4 +42,18 @@ public void setModel(String model) {
...
+
+    @Override
+    public String toString() {
+        return "id = " + id
+                + ", model: " + model
+                + ", manufacturer - " + manufacturer
+                + ", driveList: " + driverList.toString();
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/model/Car.java:57]:
```suggestion
                + ", driverList: " + driverList.toString();
```

**Review by KseniiaMakarova — APPROVED** [review]
fix comments before merging

**Pull request merged**